### PR TITLE
Feature/dac adc diagnostics

### PR DESCRIPTION
**Pull request by dan-fern**
Prior to this pull request, `throttle_control_module.ino` and `steering_control_module.ino` had no in-line test to compare DAC output to ADC input as it is fedback in.  After this pull request, this test is added (along with a `check_pedal_override` for throttle) with a failure threshold of ±96mV discrepancy between the DAC and ADC signals.  This threshold represents 3.84% of the possible DAC voltage range, 20x of the highest standard deviation and 2x more than the highest measurement recorded during testing.  Voltage spikes in the DAC output were noticed when testing the throttle module due to an increased draw from the brake pump.  These were tracked and safely filtered out.  This was found to only affect the throttle and not steering module, although a similar filter was added to the steering firmware.  Please see the attached spreadsheet for my findings. 

In addition to this test, the `dac_adc_diagnostics.ino` is modified to bench test some of the techniques implemented.

This PR makes progress towards the OSCC style guide but is not considered compliant.  Please do not review it on formatting grounds.  

[diags.xlsx](https://github.com/PolySync/OSCC/files/737231/diags.xlsx)

### CODE REVIEW

File: firmware/steering/kia_soul_ps/steering_control_module.ino [hunk 1]
```diff
@@ -21,7 +21,7 @@
 // Components:
```
Comment (ltbj) [inline firmware/steering/kia_soul_ps/steering_control_module.ino]: We decided to remove these style header comments in today’s meeting, please update.

File: firmware/steering/kia_soul_ps/steering_control_module.ino [hunk 2]
```diff
@@ -676,13 +790,23 @@ void loop( )
...
-            dac.outputA( torque_spoof.high );
-            dac.outputB( torque_spoof.low );
+            dac.outputA( torque_spoof.low );
+            dac.outputB( torque_spoof.high );
+
+            current_ctrl_state.test_countdown += 1;
+
+            // if DAC out and ADC in voltages differ, disable control
+            // only test every tenth lake cleaoop
+            if ( current_ctrl_state.test_countdown >= 10 ) {
```
Comment (ltbj) [inline firmware/steering/kia_soul_ps/steering_control_module.ino]: `{` should be on a newline.

File: firmware/tests/dac_adc_diagnostics/dac_adc_diagnostics.ino [hunk 3]
```diff
@@ -55,165 +55,80 @@ static void init_can ( void )
...
-}
+        spoof_a_adc_signal = analogRead( SPOOF_SIGNAL_A );
+        spoof_b_adc_signal = analogRead( SPOOF_SIGNAL_B );
 
-// to test that the signal interrupt relay functions we can blink two
-// LEDS in an alternating pattering using the the switch relay.
-void test_interrupt_relay( ) 
-{
-    Serial.flush(); 
+        spoof_a_adc_volts = ( spoof_a_adc_signal * 5.0 ) / 1023.0;
```
Comment (ltbj) [inline firmware/tests/dac_adc_diagnostics/dac_adc_diagnostics.ino]: What are the magic numbers `5` and `1023.0`? Please add an in-line comment.

File: firmware/tests/dac_adc_diagnostics/dac_adc_diagnostics.ino [hunk 4]
```diff
@@ -55,165 +55,80 @@ static void init_can ( void )
...
+        spoof_b_adc_volts = ( spoof_b_adc_signal * 5.0 ) / 1023.0;
 
-    while( !Serial.available() )
-    {
-        digitalWrite( SPOOF_ENGAGE, LOW );       // start off with the relay in it's depowered state
-        delay( 500 );                            // wait half a second
+        Serial.print( "DAC Value (A): " );
```
Comment (ltbj) [inline firmware/tests/dac_adc_diagnostics/dac_adc_diagnostics.ino]: All serial prints from here down to line 119 need to be removed. 

We decided that the firmware should not use any `Serial.print()` routines.

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 5]
```diff
@@ -18,16 +18,17 @@
 
 // Throttle control ECU firmware
 // Firmware for control of 2014 Kia Soul throttle system
-// Component
+// Component:
```
Comment (ltbj) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: Please remove this down to line 24.

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 6]
```diff
@@ -172,131 +192,177 @@ static void init_can ( void )
...
-  
+void calculate_pedal_spoof( float pedal_target, struct torque_spoof_t* spoof )
+{
+    // values calculated with min/max calibration curve and tuned for neutral
+    // balance.  DAC requires 12-bit values, (4096steps/5V = 819.2 steps/V)
+    spoof->low = 819.2 * ( 0.0004 * pedal_target + 0.366 );
+    spoof->high = 819.2 * ( 0.0008 * pedal_target + 0.732 );
+
+    // range = 300 - ~1750
+    spoof->low = constrain( spoof->low, 0, 1800 );
```
Comment (ltbj) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: Why is this value set outside the valid range?

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 7]
```diff
@@ -172,131 +192,177 @@ static void init_can ( void )
...
+    spoof->high = constrain( spoof->high, 0, 3500 );
+
 }
 
+//
+void check_pedal_override( )
+{
+    if ( ( signal_L + signal_H ) / 2 > PEDAL_THRESH )
+    {
+        disable_control( );
```
Comment (ltbj) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: Should this have a `DEBUG_PRINT()` indicating control is disabled?

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 8]
```diff
@@ -172,131 +192,177 @@ static void init_can ( void )
...
+        }
+    }
+    else
+    {
+        current_ctrl_state.override_flag.voltage = 0;
+        current_ctrl_state.override_flag.voltage_spike_a = 0;
+    }
+
+    // fail criteria. ~ ( ± 96mV )
+    if (    abs( spoof_b_adc_volts - spoof_b_dac_current_volts ) >
```
Comment (ltbj) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: Extra tab should be removed.

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 9]
```diff
@@ -306,16 +372,27 @@ static void publish_ps_ctrl_throttle_report( void )
...
+    tx_frame_ps_ctrl_throttle_report.id =
+            (uint32_t) (PS_CTRL_MSG_ID_THROTTLE_REPORT);
 
     // set DLC
-    tx_frame_ps_ctrl_throttle_report.dlc = 8; //TODO
+    tx_frame_ps_ctrl_throttle_report.dlc = 8;
 
     // set override flag
-    data->override = local_override;
+    if (    ( current_ctrl_state.override_flag.pedal == 0 ) &&
```
Comment (ltbj) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: Extra tabs should be removed.

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino [hunk 10]
```diff
@@ -420,19 +497,19 @@ static void check_rx_timeouts( void )
...
-    if( delta >= PS_CTRL_RX_WARN_TIMEOUT ) 
+    if( delta >= PS_CTRL_RX_WARN_TIMEOUT )
     {
         // disable control from the PolySync interface
-        if( controlEnabled ) 
+        if( current_ctrl_state.control_enabled == true )
         {
-            Serial.println("control disabled: timeout");
-            disableControl();
+            Serial.println( "control disabled: timeout" );
```
Comment (ltbj) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: Remove the serial print routine.

File: firmware/tests/dac_adc_diagnostics/dac_adc_diagnostics.ino (hunk 4 above)
Comment (dan-fern) [inline firmware/tests/dac_adc_diagnostics/dac_adc_diagnostics.ino]: These printouts are used to make sure the bench test is functioning properly.  This file does not get loaded on a deployed vehicle and is solely used to bench test.

File: firmware/throttle/kia_soul_ps/throttle_control_module.ino (hunk 7 above)
Comment (dan-fern) [inline firmware/throttle/kia_soul_ps/throttle_control_module.ino]: `disable_control( )` has a `DEBUG_PRINT( )` statement at line 267.

File: firmware/tests/dac_adc_diagnostics/dac_adc_diagnostics.ino (hunk 4 above)
Comment (ltbj) [inline firmware/tests/dac_adc_diagnostics/dac_adc_diagnostics.ino]: I had overlooked that this was in the tests directory, thanks for pointing it out.

File: firmware/steering/kia_soul_ps/steering_control_module.ino [hunk 11]
```diff
@@ -334,34 +343,105 @@ void disable_control()
...
+    {
+        current_ctrl_state.override_flag.voltage = 0;
+        current_ctrl_state.override_flag.voltage_spike_a = 0;
+    }
+
+    // fail criteria. ~ ( ± 96mV )
+    if ( abs( spoof_b_adc_volts - spoof_b_dac_current_volts ) >
+            VOLTAGE_THRESHOLD )
+    {
+        if ( current_ctrl_state.override_flag.voltage_spike_b == 0 )
```
Comment (LucasBuckland) [inline firmware/steering/kia_soul_ps/steering_control_module.ino]: Can we make this flag more descriptive? Like "voltage_spike_first_occurrence" or something?

File: firmware/steering/kia_soul_ps/steering_control_module.ino [hunk 12]
```diff
@@ -676,13 +787,23 @@ void loop( )
...
 
-            dac.outputA( torque_spoof.high );
-            dac.outputB( torque_spoof.low );
+            dac.outputA( torque_spoof.low );
+            dac.outputB( torque_spoof.high );
+
+            current_ctrl_state.test_countdown += 1;
+
+            // if DAC out and ADC in voltages differ, disable control
+            // only test every tenth lake cleaoop
```
Comment (LucasBuckland) [inline firmware/steering/kia_soul_ps/steering_control_module.ino]: Spelling errors here.

File: firmware/throttle/kia_soul_ps/current_control_state.h [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+ } override_flags;
+
+typedef struct
+{
+    //
+    //
+    bool control_enabled; /* Is control currently enabled flag */
+	//
+	//
+	bool emergency_stop; /* Emergency stop has been acitivated by higher level controller */
```
Comment (LucasBuckland) [inline firmware/throttle/kia_soul_ps/current_control_state.h]: Extra tab here it looks like.

File: firmware/steering/kia_soul_ps/steering_control_module.ino (hunk 11 above)
Comment (LucasBuckland) [inline firmware/steering/kia_soul_ps/steering_control_module.ino]: Actually, change my mind, a and b pins make sense. Don't think there's a need to change.

**Comment by dan-fern** [comment]
@petebrink, please review the attached spreadsheet with my findings and contact me with any questions.

**Pull request merged**